presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1337,6 +1339,12 @@ else if (criteria instanceof JoinOn) {
                     // coerce null to boolean
                     analysis.addCoercion(expression, BOOLEAN, false);
                 }
+                if (expression instanceof LogicalBinaryExpression) {
+                    LogicalBinaryExpression logicalExpression = (LogicalBinaryExpression) expression;
+                    if (needsJoinWarning(logicalExpression)) {
+                        warningCollector.add(new PrestoWarning(PERFORMANCE_WARNING, "JOIN conditions with an OR may lead to a cross join with filter"));
```

Comment:
1. Make it more explicit by saying - 'can cause performance issues as it may lead to a cross join with filter'
2. See if you can include the line/column number for the OR in the message

Reply:
+1 on both. And maybe we can add something prescriptive so the user can fix their query. Something like "Consider rewriting this with a UNION..."

Reply:
Done for both!